### ISSUE TITLE
False positive in mismatching_type_param_order with hardcoded type

### ISSUE THREAD

**Issue by ecton** [rust-lang/rust-clippy#9367]
### Summary

I enabled `clippy::pedantic` for a project this morning that showed a false positive I hadn't encountered before. The minimal example is below -- the definition is skipping `B` by using `()` for `B`, but I get a warning about how `C` is out of order even though the generic `C` is in the correct position.

### Lint Name

mismatching_type_param_order

### Reproducer

I tried this code:

```rust
#![warn(clippy::pedantic)]

struct Foo<A, B, C> {
    _a: A,
    _b: B,
    _c: C,
}

impl<A, C> std::fmt::Debug for Foo<A, (), C> {
    fn fmt(&self, _f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        todo!()
    }
}
```

I saw this happen:

```
warning: `Foo` has a similarly named generic type parameter `C` in its declaration, but in a different order
 --> benches/durable-writes.rs:9:43
  |
9 | impl<A, C> std::fmt::Debug for Foo<A, (), C> {
  |                                           ^
  |
note: the lint level is defined here
 --> benches/durable-writes.rs:1:9
  |
1 | #![warn(clippy::pedantic)]
  |         ^^^^^^^^^^^^^^^^
  = note: `#[warn(clippy::mismatching_type_param_order)]` implied by `#[warn(clippy::pedantic)]`
  = help: try `B`, or a name that does not conflict with `Foo`'s generic params
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#mismatching_type_param_order
```

I expected to see this happen:

No warnings.

### Version

```text
rustc 1.63.0 (4b91a6ea7 2022-08-08)
binary: rustc
commit-hash: 4b91a6ea7258a947e59c6522cd5898e7c0a6a88f
commit-date: 2022-08-08
host: x86_64-unknown-linux-gnu
release: 1.63.0
LLVM version: 14.0.5
```

**Comment by y21** [comment]
This issue seems to be fixed. Reproducer doesn't lint anymore: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=e0eec05027f54a2d8976fee42ff54938
